FungibleToken Switchboard DOCS (#74)
onflow/flow-ft: README.md

8. Use the `create_minter.cdc` transaction to create a new MintandBurn resource
   and store it in a new Admin's account.

# Fungible Token Switchboard

>  This contract, `contracts/FungibleTokenSwitchboard.cdc`, allow users to receive payments in different fungible tokens using a single `&{FungibleToken.Receiver}` placed in the standard receiver path `/public/GenericFTReceiver`.

Reviewer: ```suggestion
`FungibleTokenSwitchboard.cdc`, allows users to receive payments in different fungible tokens using a single `&{FungibleToken.Receiver}` placed in a standard receiver path `/public/GenericFTReceiver`.
```

 ## How to use it

>  Users willing to use the Fungible Token Switchboard will need to setup their accounts by creating a new `FungibleTokenSwitchboard.Switchboard` resource and saving it to their accounts. This can be acomplished by executing the transaction found in this repository `transactions/switchboard/setup_account.cdc`. This transaction will create and save a Switchboard resource to the signer's account, and it also will create the needed public capabilities to access it. After setting up their switchboard users will need to add the fungible token's vault with which they are willing to receive payments to their switchboard resource.

Reviewer: ```suggestion
 Users willing to use the Fungible Token Switchboard will need to setup their accounts by creating a new `FungibleTokenSwitchboard.Switchboard` resource and saving it to their accounts at the `FungibleTokenSwitchboard.StoragePath` path.
 
 This can be acomplished by executing the transaction found in this repository `transactions/switchboard/setup_account.cdc`. This transaction will create and save a Switchboard resource to the signer's account,
 and it also will create the needed public capabilities to access it. After setting up their switchboard, in order to make it support receiving a certain token, users will need to add the desired token's receiver capability to their switchboard resource.
```

 
 ## Adding a new vault to the switchboard

>  When a user wants to receive a new fungible token through their switchboard they will need to add a new public capability linked to said FT to their swichboard resource. This could be acomplished in two different ways:

Reviewer: ```suggestion
 When a user wants to receive a new fungible token through their switchboard, they will need to add a new public capability linked to said FT to their switchboard resource. This can be accomplished in two different ways:
```

    transaction {
        let exampleTokenVaultCapabilty: Capability<&{FungibleToken.Receiver}>
        let switchboardRef:  &FungibleTokenSwitchboard.Switchboard

        prepare(signer: AuthAccount) {
          // Get the example token vault capability from the signer's account
          self.exampleTokenVaultCapabilty = 
            signer.getCapability<&{FungibleToken.Receiver}>
                                    (ExampleToken.ReceiverPublicPath)
          // Get a reference to the signers switchboard
          self.switchboardRef = signer.borrow<&FungibleTokenSwitchboard.Switchboard>
            (from: FungibleTokenSwitchboard.StoragePath) 
              ?? panic("Could not borrow reference to switchboard")
        }

        execute {
          // Add the capability to the switchboard using addNewVault method
          self.switchboardRef.addNewVault(capability: self.exampleTokenVaultCapabilty)
        }
    }
    ```
    This function will panic if is not possible to `.borrow()` a reference to a `&{FungibleToken.Receiver}` from the passed capability. It will also panic if there is already a capability stored for the same `Type` of resource exposed by the capability.

>  1. Adding one or more capabilities using the paths where they are stored using `addNewVaultsByPath(paths: [PublicPath], address: Address)`

Reviewer: ```suggestion
 2. Adding one or more capabilities using the paths where they are stored using `addNewVaultsByPath(paths: [PublicPath], address: Address)`
```

>     * When using this method, an array of `PublicPath` objects should be pass along with the `Address` of the account from where the vaults' capabilities should be retrieved.

Reviewer: ```suggestion
    * When using this method, an array of `PublicPath` objects should be passed along with the `Address` of the account from where the vaults' capabilities should be retrieved.
```

          // Get the example token vault path from the contract
          self.exampleTokenVaultPath = ExampleToken.ReceiverPublicPath
          // And store it in the array of public paths that will be passed to the
          // switchboard method
          self.vaultPaths = []
          self.vaultPaths.append(self.exampleTokenVaultPath)
          // Get a reference to the signers switchboard
          self.switchboardRef = signer.borrow<&FungibleTokenSwitchboard.Switchboard>
            (from: FungibleTokenSwitchboard.StoragePath) 
              ?? panic("Could not borrow reference to switchboard")
        }

        execute {
          // Add the capability to the switchboard using addNewVault method
          self.switchboardRef.addNewVaultsByPath(paths: self.vaultPaths, 
                                                        address: address)
        }
    }
    ```
    This function won't panic, instead it will just not add to the `@Switchboard` any capability which can not be retrieved from any of the provided `PublicPath`s. It will also ignore any type of `&{FungibleToken.Receiver}` that is already present on the `@Switchboard`

 ## Removing a vault from the switchboard

>  If a user no longer wants to be able to receive deposits from a certain FT, or if they want to update the provided capability for one of them, they will need to remove the vault from the switchboard. This can me acomplish by using `removeVault(capability: Capability<&{FungibleToken.Receiver}>)`. As for adding new vaults it is required to first retrieve the vault's capability from the user account and passing it as a parameter. This can be observed in the template transaction `transactions/switchboard/remove_vault_capability.cdc`:

Reviewer: ```suggestion
 If a user no longer wants to be able to receive deposits from a certain FT, or if they want to update the provided capability for one of them, they will need to remove the vault from the switchboard.
 This can be accomplished by using `removeVault(capability: Capability<&{FungibleToken.Receiver}>)`. 
This can be observed in the template transaction `transactions/switchboard/remove_vault_capability.cdc`:
```

    prepare(signer: AuthAccount) {
      // Get the example token vault capability from the signer's account
      self.exampleTokenVaultCapabilty = signer.getCapability
                    <&{FungibleToken.Receiver}>(ExampleToken.ReceiverPublicPath)
      // Get a reference to the signers switchboard  
      self.switchboardRef = signer.borrow<&FungibleTokenSwitchboard.Switchboard>
        (from: FungibleTokenSwitchboard.StoragePath) 
          ?? panic("Could not borrow reference to switchboard")

    }

    execute {
      // Remove the capability from the switchboard using the 
      // removeVault method
      self.switchboardRef.removeVault(capability: self.exampleTokenVaultCapabilty)
    }
 }
 ```
 This function will panic if is not possible to `.borrow()` a reference to a `&{FungibleToken.Receiver}` from the passed capability.

 ## Transfering tokens through the switchboard
 The Fungible Token Switchboad provides two different ways of depositing tokens to it, using the `deposit(from: @FungibleToken.Vault)` method enforced by the `{FungibleToken.Receiver}` or using the `safeDeposit(from: @FungibleToken.Vault): @FungibleToken`:

>  1. Using the first method will be just the same as depositing to `&{FungibleToken.Receiver}`. The path for the Switchboard receiver is defined in `FungibleTokenSwitchboard.ReceiverPublicPath`, the generic receiver path `/public/GenericFTReceiver` that can also be obtained from the NFT MetadataViews contract. An example of how to do this could be find in the transaction template on this repo `transactions/switchboard/transfer_tokens.cdc`

Reviewer: ```suggestion
 1. Using the first method will be just the same as depositing to `&{FungibleToken.Receiver}`. The path for the Switchboard receiver is defined in `FungibleTokenSwitchboard.ReceiverPublicPath`,
 the generic receiver path `/public/GenericFTReceiver` that can also be obtained from the NFT MetadataViews contract.
 An example of how to do this can be found in the transaction template on this repo `transactions/switchboard/transfer_tokens.cdc`
```
